### PR TITLE
Evm support

### PR DESCRIPTION
**Pull request by yaroslavyaroslav**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/analysis/p/analysis_evm_cs.c [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+	if (!strcmp(a->cpu, "v9")) {
+		mode |= CS_MODE_V9;
+	}
+	if (mode != omode) {
+		cs_close(&handle);
+		handle = 0;
+		omode = mode;
+	}
+	if (handle == 0) {
+		ret = cs_open(CS_ARCH_SPARC, mode, &handle);
```
Comment (XVilka) [inline librz/analysis/p/analysis_evm_cs.c]: Defininetly not `SPARC`

File: librz/analysis/p/analysis_evm_cs.c [hunk 2]
```diff
@@ -0,0 +1,419 @@
+// SPDX-FileCopyrightText: 2014-2017 pancake <pancake@nopcode.org>
```
Comment (XVilka) [inline librz/analysis/p/analysis_evm_cs.c]: Put yourself and use 2023

File: librz/analysis/p/analysis_evm_cs.c [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+#include <capstone/sparc.h>
+
+#if CS_API_MAJOR < 5
+#error Old Capstone not supported
+#endif
+
+#define INSOP(n) insn->detail->sparc.operands[n]
+#define INSCC    insn->detail->sparc.cc
+
+static void opex(RzStrBuf *buf, csh handle, cs_insn *insn) {
```
Comment (XVilka) [inline librz/analysis/p/analysis_evm_cs.c]: Can remove this handler completely

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/p/analysis_evm_cs.c (hunk 1 above)
Comment (yaroslavyaroslav) [reply librz/analysis/p/analysis_evm_cs.c]: The first goal is to just register that arch and anal plugins in the core. The whole logic is completely untouched yet.

**Comment by yaroslavyaroslav** [comment]
@XVilka I've posted the questions at mattermost, but let me duplicate it here.

How should i register both arch and analysis plugin in the system? At the current state none of them are available in anywhere. All of the follow are failing:
- `rz-asm -L | grep evm`
- `rizin -a evm[.cs]`

They also doesn't appears on `mason setup --wipe …` in the arch and analysis plugins list.

UPD: Nah, found it.

File: librz/asm/p/asm_evm_cs.c [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	return ret;
+}
+
+RzAsmPlugin rz_asm_plugin_evm_cs = {
+	.name = "evm.cs",
+	.desc = "Capstone EVM disassembler",
+	.license = "BSD",
+	.arch = "evm",
+	.cpus = "v9",
+	.bits = 256,
```
Comment (yaroslavyaroslav) [inline librz/asm/p/asm_evm_cs.c]: Failed to support that value. EVM word length and integer is 256 at most.

**Comment by XVilka** [comment]
It's probably better to wait first once auto-sync work is merged first, since we will update capstone after that:
- https://github.com/rizinorg/rizin/pull/3463
- https://github.com/rizinorg/rizin/pull/3399
- https://github.com/rizinorg/rizin/pull/3301

**Comment by XVilka** [comment]
It will also need tests

**Comment by yaroslavyaroslav** [comment]
@XVilka I consider to not use capstone one as the back for EVM since last time I've checked it had outdated opcodes set.

**Comment by Rot127** [comment]
@yaroslavyaroslav Is [this](https://github.com/etclabscore/evm_llvm/tree/EVM/lib/Target/EVM) LLVM fork up to date? Because in this case I would also recommend to wait until I finish the docs for auto-sync (in the next two or three days) and use `auto-sync` on this fork.
TriCore (https://github.com/capstone-engine/capstone/pull/1973) did the same recently.

**Comment by yaroslavyaroslav** [comment]
@Rot127 At a glance view yes it is. Thank you for pointing me out to that one. And thank you, I'll be waiting till your work will be done.

**Comment by Rot127** [comment]
@yaroslavyaroslav Just pushed the documentation (https://github.com/capstone-engine/capstone/pull/1949/commits/478041681a92d6a404951c4523076140dab918da). You are very much welcome to give feedback or even start implementing.

**Comment by Rot127** [comment]
@yaroslavyaroslav Can I add you to the list here: https://github.com/capstone-engine/capstone/issues/2015?

**Comment by yaroslavyaroslav** [comment]
> @yaroslavyaroslav Can I add you to the list here: https://github.com/capstone-engine/capstone/issues/2015?

Likely yes, but I plan to get back to that work somewhere around July. So I hope there's no rush there.

**Comment by Rot127** [comment]
Alright. Then I leave you off the list for now. Please just don't forget to ping us when you start. Just in case someone else already started.

**Comment by yaroslavyaroslav** [comment]
Sorry folks, I've completely out of this now, so there's no chance that I finish this PR.

**Pull request closed without merging**